src/client.ts

```diff
@@ -165,95 +149,151 @@ export class SubscriptionClient {
+      this.unsubscribe(parseInt(subId, 10));
     });
   }
 
-  // send message, or queue it if connection is not open
-  private sendMessage(message: Object) {
-    switch (this.client.readyState) {
+  private executeRequest(options: RequestOptions, handler: (error: Error[], result?: any) => void): number {
+    const { query, variables, operationName } = options;
+
+    if (!query) {
+      throw new Error('Must provide a `query`.');
+    }
+
+    if (!handler) {
+      throw new Error('Must provide an `handler`.');
+    }
+
+    if (
+      ( !isString(query) && !getOperationAST(query, operationName)) ||
+      ( operationName && !isString(operationName)) ||
+      ( variables && !isObject(variables))
+    ) {
+      throw new Error('Incorrect option types. `query` must be a string or a document,' +
+        '`operationName` must be a string, and `variables` must be an object.');
+    }
+
+    const reqId = this.generateRequestId();
+    this.requests[reqId] = { options, handler };
+    this.sendMessage(reqId, MessageTypes.GQL_START, options);
+
+    return reqId;
+  }
+
+  private buildMessage(id: number, type: string, payload: any) {
+    const payloadToReturn = payload && payload.query ?
+      {
+        ...payload,
+        query: typeof payload.query === 'string' ? payload.query : print(payload.query),
+      } :
+      payload;
+
+    return {
+      id,
+      type,
+      payload: payloadToReturn,
+    };
+  }
 
+  // ensure we have an array of errors
+  private formatErrors(errors: any) {
+    if (Array.isArray(errors)) {
+      return errors;
+    }
+    if (errors && errors.message) {
+      return [errors];
+    }
+    return [{ message: 'Unknown error' }];
+  }
+
+  private sendMessage(id: number, type: string, payload: any) {
+    this.sendMessageRaw(this.buildMessage(id, type, payload));
+  }
+
+  // send message, or queue it if connection is not open
+  private sendMessageRaw(message: Object) {
+    switch (this.status) {
       case this.client.OPEN:
-        // TODO: throw error if message isn't json serializable?
-        this.client.send(JSON.stringify(message));
+        let serializedMessage: string = JSON.stringify(message);
+        let parsedMessage: any;
+        try {
+          parsedMessage = JSON.parse(serializedMessage);
+        } catch (e) {
+          throw new Error(`Message must be JSON-serializable. Got: ${message}`);
+        }
 
+        this.client.send(serializedMessage);
         break;
       case this.client.CONNECTING:
         this.unsentMessagesQueue.push(message);
 
         break;
       case this.client.CLOSING:
       case this.client.CLOSED:
+        break;
       default:
         if (!this.reconnecting) {
           throw new Error('Client is not connected to a websocket.');
         }
     }
   }
 
-  private generateSubscriptionId() {
-    const id = this.maxId;
-    this.maxId += 1;
+  private generateRequestId() {
+    const id = this.nextRequestId;
+    this.nextRequestId += 1;
```

Comment:
If we do 10 requests per second, this will overflow after 28 million years. I'm fine with that 🙂